File: crates/mithril-oxide-sys/cpp/IR/Location.cpp
#include "Location.hpp"

#include <memory>

#include <mlir/IR/Location.h>
#include <mlir/IR/MLIRContext.h>
#include <mlir/IR/Attributes.h>
#include <mlir/IR/BuiltinAttributes.h>


namespace mithril_oxide_sys {

// TODO: Proxies for FusedLoc.
// TODO: Proxies for Location.
// TODO: Proxies for NameLoc.
// TODO: Proxies for OpaqueLoc.

const void* UnknownLoc_get(MLIRContext &ctx)
{
    return UnknownLoc::get(&ctx).getAsOpaquePointer();
}

const void* FileLineColLoc_get(const void* filename_ptr, unsigned line, unsigned column)
{
    auto filename = mlir::StringAttr::getFromOpaquePointer(filename_ptr);
    return FileLineColLoc::get(filename, line, column).getAsOpaquePointer();
}

const void* CallSiteLoc_get(const void* callee_ptr, const void* caller_ptr)
{
    return CallSiteLoc::get(
        Location::getFromOpaquePointer(callee_ptr),
        Location::getFromOpaquePointer(caller_ptr)
    ).getAsOpaquePointer();
}

} // namespace mithril_oxide_sys

File: crates/mithril-oxide-sys/src/IR/Location.rs
pub use self::ffi::{CallSiteLoc_get, FileLineColLoc_get, UnknownLoc_get};

#[cxx::bridge]
pub(crate) mod ffi {
    #[namespace = "mlir"]
    unsafe extern "C++" {
        include!("mithril-oxide-sys/cpp/IR/Location.hpp");

        type MLIRContext = crate::IR::MLIRContext::MLIRContext;
    }

    #[namespace = "mithril_oxide_sys"]
    unsafe extern "C++" {
        include!("mithril-oxide-sys/cpp/IR/Location.hpp");

        type c_void = crate::c_void;

        #[must_use]
        pub unsafe fn UnknownLoc_get(ctx: Pin<&mut MLIRContext>) -> *const c_void;

        // filename - stringattr
        #[must_use]
        pub unsafe fn FileLineColLoc_get(
            filename: *const c_void,
            line: u32,
            column: u32,
        ) -> *const c_void;

        // callee - location
        // caller - location
        #[must_use]
        #[allow(clippy::similar_names)]
        pub unsafe fn CallSiteLoc_get(
            callee_ptr: *const c_void,
            caller_ptr: *const c_void,
        ) -> *const c_void;
    }
}
